cmd/clusterctl/cmd/backup.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"github.com/spf13/cobra"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/client"
+)
+
+type backupOptions struct {
+	fromKubeconfig        string
+	fromKubeconfigContext string
+	namespace             string
+	directory             string
+}
+
+var buo = &backupOptions{}
+
+var backupCmd = &cobra.Command{
+	Use:   "backup",
+	Short: "Backup Cluster API objects and all dependencies from a management cluster.",
+	Long: LongDesc(`
+		Backup Cluster API objects and all dependencies from a management cluster.
+
+		TODO`),
+
+	Example: Examples(`
+		Backup Cluster API objects and all dependencies from a management cluster.
+		clusterctl backup --directory-kubeconfig=/tmp/backup-directory`),
+	Args: cobra.NoArgs,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return runBackup()
+	},
+}
+
+func init() {
+	backupCmd.Flags().StringVar(&buo.fromKubeconfig, "kubeconfig", "",
+		"Path to the kubeconfig file for the source management cluster to backup. If unspecified, default discovery rules apply.")
+	backupCmd.Flags().StringVar(&buo.fromKubeconfigContext, "kubeconfig-context", "",
+		"Context to be used within the kubeconfig file for the source management cluster. If empty, current context will be used.")
+	backupCmd.Flags().StringVarP(&buo.namespace, "namespace", "n", "",
+		"The namespace where the workload cluster is hosted. If unspecified, the current context's namespace is used.")
+	backupCmd.Flags().StringVar(&buo.directory, "directory", "",
```

Comment:
I didn't make the directory flag required since the `setObjectsDir` func will default to the Cluster API default config folder under the `objects` directory if no directory is provided. This was implemented [here](https://github.com/kubernetes-sigs/cluster-api/pull/4808/files#diff-972d17a2075f3710e12506a9a27cb2b1a0f67b6b69dbc7eb342d1adff36a7d2eR197-R213). If we expect the user to provide a directory no matter what, then maybe we remove this "default" logic from this func and cobra will error early if the flag is not provided

Reply:
I think removing the defaulting logic somehow makes sense, especially given that restore takes every yaml in a folder.
I'm +1 to ask the user to be explicit about the directory to use when doing backup/restore

Reply:
:+1: Removed glob and now require the `directory` flag for both backup and restore